include/beast/websocket/impl/ping.ipp

```diff
@@ -146,24 +146,31 @@ operator()(error_code ec, bool again)
                         boost::asio::error::operation_aborted));
                 return;
             }
-            // fall through
+            d.ws.wr_block_ = &d;
+            // [[fallthrough]]
 
         case 1:
             // send ping frame
             d.state = 99;
-            BOOST_ASSERT(! d.ws.wr_block_);
-            d.ws.wr_block_ = &d;
             boost::asio::async_write(d.ws.stream_,
                 d.fb.data(), std::move(*this));
             return;
 
         case 2:
+            BOOST_ASSERT(! d.ws.wr_block_);
```

Comment:
`upcall` is a final state, once we perform the upcall (the call to the final completion handler) the composed operation is done and gets destroyed. When `wr_block_` is set and we suspend, it is the other blocking operation which clears `wr_block_` before the upcall (or continuing the composed operation such as in the case of `write_op`).